File: Python/13549_숨바꼭질3.py
import sys

if __name__ == '__main__':
	N, K = list(map(int,sys.stdin.readline().split()))
	
	q = [[N,0]] # bfs를 진행할 큐 ** priority queue를 이용해서 풀어봐야 할 듯...
	visit = [False]*100001 # 방문 여부 확인 배열
	visit[N] = True
	
	result = 0 # 최종적으로 걸린 시간
	while q:
		x, time = q.pop(0)
		
		if x == K: # 수빈이가 동생 위치에 도달하였다면
			result = time # 시간 저장 후 break
			break
		
		if 2*x <= 100000 and visit[2*x] == False: # 순간이동 먼저 넣어주어야 함 -> 시간이 덜 걸리기 때문
			q.append([2*x,time])
			visit[2*x] = True
		if x-1 >= 0 and visit[x-1] == False: # 한 칸 앞으로 전진
			q.append([x-1,time+1])
			visit[x-1] = True
		if x+1 <= 100000 and visit[x+1] == False: # 한 칸 뒤로 후진
			q.append([x+1,time+1])
			visit[x+1] = True
		
	print(result)

File: C++/13549_숨바꼭질3.cpp
#include <iostream>
#include <queue>
using namespace std;

struct Pos{
	int x, time;
};

int main() {
	int N, K;
	cin >> N >> K;
	
	bool visit[100001] = {0,}; // 수빈이가 현재 위치에 방문했는지의 여부를 담을 배열
	visit[N] = true;
	
	queue<Pos> q;
	q.push({N,0}); // bfs 위한 queue, 수빈이의 초기 위치로 초기화(시간은 0초)  *** priority queue를 이용해야 할 것 같음..
	
	int result = 0; // 동생을 찾는데 걸리는 최종 시간
	while(!q.empty()){
		Pos p = q.front();
		q.pop();
		
		if(p.x == K){ // 동생 위치에 도달(동생을 찾았다면)
			result = p.time; // 시간 저장 후 while문 탈출
			break;
		}
		
		if(2*p.x <= 100001 && visit[2*p.x] == false){ // 순간이동할 경우(시간이 증가하지 않기 때문에 가장 먼저 확인하여 넣어 주어야 함)
			q.push({2*p.x,p.time});
			visit[2*p.x] = true;
		}
		if(p.x-1 >= 0 && visit[p.x-1] == false){ // 뒤로 한 칸 이동
			q.push({p.x-1,p.time+1});
			visit[p.x-1] = true;
		}
		if(p.x+1 <= 100001 && visit[p.x+1] == false){ // 앞으로 한 칸 이동
			q.push({p.x+1,p.time+1});
			visit[p.x+1] = true;
		}
	}
	
	cout << result;
	
	return 0;
}
